[22기 최재상] 불필요한 consturctor, state, 데미데이터 삭제 및 전체파일 구조분해 할당

Higher77 opened the pull request:
>wecode PR message template 및 체크 리스트 입니다. 
아래 사항들을 전부 작성/체크 하시고 PR 해주세요!

## 수정 사항 간략한 한줄 요약
(여기에 수정 사항에 대한 간략한 한줄 요약/제목 작성해 주세요.)
- 불필요한 constructor, state, 더미데이터 삭제 및 전체파일 구조분해 할당
		 
## 수정 사항들 자세한 내용
(여기에 수정 사항에 대한 자세한 내용을 작성해 주세요.)
- EvaluationContainer.js , SimilarItem.js, WantEatContainer.js 의 constructor 삭제
- Slider.js의 State에 저장된 this.props.subImage삭제
- eachDatalist, categoryDatalist 더미 데이터 삭제 
- 전체파일 구조분해 할당

## 기타 질문 및 특이 사항
(궁금한 사항들, 하면서 느낀 점들 공유해주실 것이 있으면 자유롭게 작성해 주세요.)

## 체크 리스트 (아래 사항들이 전부 체크되어야만 merge가 됩니다!)
- [x] 필요한 test들을 완료하였고 기능이 제대로 실행되는지 확인 하였습니다.
- [x] Wecode의 코드 스타일 가이드에 맞추어 코드를 작성 하였습니다.
- [x] 제가 의도한 파일들과 수정 사항들만 커밋이 된 것을 확인 하였습니다.
- [x] 본 수정 사항들을 팀원들과 사전에 상의하였고 팀원들 모두 해당 PR에 대하여 알고 있습니다.

Commit:
[Add] Detail 레이아웃 1차 제출본
-카테고리 이미지 블러효과 아직 못 넣음
-상품의 다른이미지 블러효과 아직 못 넣음
-비슷한 상품에 간략한 상품설명 못 달음

Tonyk0901 commented on src/Pages/Detail/Detail.js:
```diff
@@ -1,7 +1,107 @@
+            <div className="anotherItemContainer">
+              <h2>상품의 다른이미지</h2>
+              <div className="anotherItemList">
+                <div className="anotherItem">
+                  <img src="https://images.unsplash.com/photo-1622708862830-a026e3ef60bd?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=630&q=80" />
+                </div>
+                <div className="anotherItem">
+                  <img src="https://images.unsplash.com/photo-1622708862830-a026e3ef60bd?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=630&q=80" />
+                </div>
+                <div className="anotherItem">
+                  <img src="https://images.unsplash.com/photo-1622708862830-a026e3ef60bd?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=630&q=80" />
+                </div>
+                <div className="anotherItem">
+                  <img src="https://images.unsplash.com/photo-1622708862830-a026e3ef60bd?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=630&q=80" />
+                </div>
```
- 이 부분은 후에 백엔드에서 받아오는 데이터를 사용해서 구현되실 것 같은데, 그때는 이렇게 이미지 url 을 직접적으로 넣어서 사용하실 수 없을 겁니다.
- 이미지 url 부분을 배열로 만드신 뒤, map 함수를 사용하여 UI 를 구현하는 방식으로 하셔야 이후에 백엔드와 맞춰보실때 좀 더 수월하게 진행될 수 있을 것 같습니다.

Tonyk0901 commented on src/Pages/Detail/Detail.js:
```diff
@@ -1,7 +1,107 @@
+            <div className="similarItemContainer">
+              <h2>비슷한 상품</h2>
+              <div className="similarItemList">
+                <div className="similarItem">
+                  <img src="https://images.unsplash.com/photo-1622708862830-a026e3ef60bd?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=630&q=80" />
+                </div>
+                <div className="similarItem">
+                  <img src="https://images.unsplash.com/photo-1622708862830-a026e3ef60bd?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=630&q=80" />
+                </div>
+                <div className="similarItem">
+                  <img src="https://images.unsplash.com/photo-1622708862830-a026e3ef60bd?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=630&q=80" />
+                </div>
+                <div className="similarItem">
+                  <img src="https://images.unsplash.com/photo-1622708862830-a026e3ef60bd?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=630&q=80" />
+                </div>
+                <div className="similarItem">
+                  <img src="https://images.unsplash.com/photo-1622708862830-a026e3ef60bd?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=630&q=80" />
+                </div>
+                <div className="similarItem">
+                  <img src="https://images.unsplash.com/photo-1622708862830-a026e3ef60bd?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=630&q=80" />
+                </div>
```
- 이 부분도 반복되는 구조네요! 달라지는 부분을 목업 데이터로 구성해주신 뒤에 map 함수를 사용해서 구현해주세요!

Higher77 replied:
넵 하위컴포넌트로 분리했습니다!

Tonyk0901 reviewed (commented):
고생하셨습니다 재상님-! 크게 문제될 것 같아 보이는 부분은 없네요! 코멘트 확인해주세요!

- detail 에서 main 부분은 기능을 추가하게 되면 map 을 사용하고 여러 로직이 들어가는 등 자체적으로 복잡해질 것 같습니다. 그래서 이 부분은 따로 하위 컴포넌트로 분리하시는 것도 좋을 것 같습니다.

Commit:
[Add] 레이아웃 완성 및 케러셀 구현 완료
-레이아웃 구성을 완료했습니다.
-이미지 케러셀을 완성했습니다.
-CategoryImage만 컴포넌트화 했습니다.

Commit:
[Add] 컴포넌트화 완료 및 백엔드 데이터 통신 성공!

Commit:
[Add]슬라이더 Next,Prev버튼 토글기능 구현, CSS 기능 추가
-슬라이더: 토글기능을 구현해 처음엔 Prev버튼이 안나오게하고 마지막엔 Next버튼이 안나오게 함
-비슷한 상품에 마우스 호버시 애니메이션 추가
-카테고리 이미지 CSS완성

Commit:
[Add] 백엔드로부터 받은 데이터 적용중
-AnotherItem 컴포넌트를 map함수에 적용하기 전 백업커밋입니다.

Tonyk0901 commented on src/Pages/Detail/AnotherItem/AnotherItem.js:
```diff
@@ -0,0 +1,16 @@
+export default class AnotherItem extends Component {
+  constructor(props) {
+    super(props);
+  }
```
- props 는 자동으로 들어가기 때문에, 이렇게 해주실 필요가 없습니다.
- constructor 에서 props 를 표기해주는 이유는, **constructor 안에서** props 에 접근하기 위함입니다. 예를 들어 state 에 props 로 받은 값을 초기 값으로 넣는 경우가 있습니다.
```js
constructor(props) {
  super(props);
  this.state = {
    test: this.props.testProps, // 매개변수와 super 에 props 를 명시하지 않으면 props 를 읽을 수 없다고 에러남.
  }
}
```

Tonyk0901 commented on src/Pages/Detail/CategoryImage/CategoryImage.js:
```diff
@@ -0,0 +1,25 @@
+export default class CategoryImage extends Component {
+  constructor(props) {
+    super(props);
+  }
```
- 불필요한 코드입니다! 지워주세요!

Commit:
[Add] AnotherItem컴포넌트 Map함수 적용

Commit:
[Merge] Master branch로부터 Merge 진행

Tonyk0901 commented on src/Pages/Detail/Detail.js:
```diff
@@ -1,7 +1,175 @@
+                  <AnotherItem itemRef={this.itemRef} />
+                  <AnotherItem itemRef={this.itemRef} />
+                </div>
+              </div>
```
- 슬라이더 부분은 그 자체로 이미 복잡하기 때문에, 개별 컴포넌트로 만들어서 관리해주시는게 좋을 것 같습니다.
- 슬라이더 되는 이미지를 하드코딩 하는 것이 아니라, img url 을 배열로 만들어서 관리해 주시는게 더 좋습니다. 이렇게 되면, 예를 들어, `sliderElement.getElementsByClassName('anotherItem').length`  와 같이 getElementsByClassName 메서드를 사용해서 길이를 계산하는 등의 부가적인 로직이 필요 없어 집니다. 왜냐하면 이미지 url 이 담긴 배열의 길이 정보를 가지고 length 를 바로 계산하면 되기 때문이죠!

Tonyk0901 commented on src/Pages/Detail/Detail.js:
```diff
@@ -1,7 +1,175 @@
+  componentDidMount = () => {
+    const sliderElement = this.sliderRef.current;
+    const itemElement = this.itemRef.current;
+    const itemWidth = Math.ceil(itemElement.getBoundingClientRect().width);
```
- itemElement 변수를 거치는 것 보다, 그냥 itemWidth 를 ref 를 사용해서 바로 계산하는게 더 좋아보입니다!
```suggestion
    const itemWidth = Math.ceil(this.itemRef.current.getBoundingClientRect().width);
```

Tonyk0901 commented on src/Pages/Detail/Detail.js:
```diff
@@ -1,7 +1,175 @@
+  // };
+
+  componentDidMount = () => {
+    const sliderElement = this.sliderRef.current;
```
- 슬라이더 안에 들어가는 이미지들을 배열 state로 관리한다면, sliderElement 라고 요소에 접근해서 anotherItem 클래스 네임을 가진 요소들을 전부 가져온 후 길이를 계산할 필요가 없습니다.

Higher77 replied:
말씀해주신대로 리팩토리 진행했습니다! 언제나 감사합니다!!

Tonyk0901 reviewed (commented):
코멘트 확인해주세요! 슬라이더 잘 구현해주셨는데, 몇가지 수정해주시면 더 좋을 것 같습니다.

1. 위치 정보 자체를 state 로 만들어서 저장하고 계신데, 이것 보다 카드의 인덱스를 state 로 만드신 뒤에, 해당 state 를 사용해서 위치를 계산하는 방식으로 구현을 해주세요! 슬라이더의 위치 정보는 용도가 단 하나 이지만, 인덱스 정보는 여러가지 용도로 사용될 수 있어 확장성이 더 좋습니다.

2. Detail.js 에는 여러가지 로직들이 들어올 예정인데, 슬라이더 관련한 코드들도 같이 섞여있어 이후에 매우 복잡해 질 것 같습니다. 이런 경우에, Slider 부분을 컴포넌트로 만들어서 분리해 주시는게 필요할 것 같습니다.

어느 정도 보여 드리자면 아래와 같이 코드가 구성될 것 같습니다. 어떤 느낌인지 **참고만** 해주세요!

```jsx
class Slider extends component {
  state = {
    currentIndex: 0,
  }
  slider = React.createRef();
  itemRef = React.createRef();

  
  handleNextSlider = () => {
    const { currentIndex } = this.state;
    const { imageData } = this.props;
    const itemWidth = Math.ceil(
      this.itemRef.current.getBoundingClientRect().width
    );

    if (currentIndex === imageData.length - 4) return;

    this.setState(
      {
        currentIndex: currentIndex + 1;
      },
      () => {
        slider.current.style.transform = `translate(-${currentIndex * itemWidth}px)`;
      }
    );
  };
  
  // 이외의 슬라이더 관련 로직이 더 들어오겠죠?
  
  render() {
    return (
      <ul className='sliderContainer' ref={slider}>
        {this.props.imageData.map((image, index) => {
          return (
             <AnotherItem ref={itemRef} className='sliderImage' key={index}>{image}</li>
          )
        })}
      </ul>
    )
  }
}
```

Commit:
[Add] Refactory 슬라이더 진행
-슬라이더의 state를 위치정보가 아닌 인덱스로 변경중입니다.
-백업데이터 입니다.

Commit:
[Refactor] 슬라이더 위치정보를 인덱스로 변경
-state에 저장된 슬라이더 위치정보를 인덱스로 바꿔 확장성을 더 좋게 만들었습니다.

Commit:
[Refactor] Slider 컴포넌트 분리

Commit:
[Add] SimilarItem컴포넌트 key값 추가

Tonyk0901 commented on src/Pages/Detail/EvaluationContainer/EvaluationContainer.js:
```diff
@@ -0,0 +1,16 @@
+import './EvaluationContainer.scss';
+
+export default class EvaluationContainer extends Component {
+  constructor(props) {
+    super(props);
+  }
```
- 불필요한 컨스트럭터 입니다! 지워주세요!

Tonyk0901 commented on src/Pages/Detail/SimilarItem/SimilarItem.js:
```diff
@@ -0,0 +1,21 @@
+import './SimilarItem.scss';
+
+export default class SimilarItem extends Component {
+  constructor(props) {
+    super(props);
+  }
```
- 지워주세요!

Tonyk0901 commented on src/Pages/Detail/WantEatContainer/WantEatContainer.js:
```diff
@@ -0,0 +1,22 @@
+import './WantEatContainer.scss';
+
+export default class WantEatContainer extends Component {
+  constructor(props) {
+    super(props);
+  }
```
- 지워주세요!

Tonyk0901 commented on src/Pages/Detail/Slider/Slider.js:
```diff
@@ -0,0 +1,105 @@
+    this.itemRef = React.createRef();
+    this.state = {
+      currentIndex: 0,
+      subImage: this.props.subImage,
```
- props 로 받은 이미지 배열을 다시 state 로 저장해서 관리할 필요가 없어 보입니다! 수정이 필요하지 않고 그냥 읽기만 하면 되기 때문이죠!

Tonyk0901 commented on src/Pages/Detail/Detail.js:
```diff
@@ -1,7 +1,103 @@
+    this.sliderRef = React.createRef();
+    this.itemRef = React.createRef();
+    this.state = {
+      eachDatalist: {
+        sub_image_url: [1, 2, 3, 4, 5, 6, 7],
+      },
+      categoryDatalist: {
+        category_image_url: [1, 2, 3, 4, 5, 6, 7, 8],
+      },
```
- 일단 더미데이터 인가요?

Higher77 replied:
네, 더미데이터입니다. 통신시작하면 각각의 데이터를 eachDataList와 categoryDataList에 저장할 생각이에요!

Tonyk0901 reviewed (commented):
슬라이더, 컴포넌트로 잘 분리시켜 주셨네요-! ✨  몇가지 더 작은 수정사항 있는데, 확인해주세요!

- 전체적으로 구조분해 할당을 잘 활용하지 않고 계신데, 구조분해할당 활용해주세요!
- Detail.js 에서 슬라이더 관련한 코드들은 지워주세요!

Commit:
[Refactor] 불필요한 constructor,State삭제

Commit:
[Refactor] 전체파일 구조분해 할당
-전체파일 구조분해 할당
-Detail.js eachDatalist,categoryDatalist  더미데이터 삭제 후 백엔드 서버로부터 데이터를 받아옴

Tonyk0901 commented:
리뷰 사항들 잘 반영해주셨네요! 크게 리뷰가 필요한 부분은 더 없는 것 같습니다-! ✨  추가적으로 기능 구현될 브랜치인가요?

Higher77 commented:
나중에 별점 컴포넌트 삽입하는거 말고는 detail 내에서 더 추가할 기능은 없습니다! merge부탁드려요!
